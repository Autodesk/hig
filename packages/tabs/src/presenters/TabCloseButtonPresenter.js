import React from "react";
import PropTypes from "prop-types";
import { CloseSUI, CloseXsUI } from "@hig/icons";
import ThemeContext from "@hig/theme-context";
import { ControlBehavior } from "@hig/behaviors";
import { cx, css } from "emotion";
import stylesheet from "./Tab.stylesheet";

export default class TabCloseButtonPresenter extends React.Component {
  static propTypes = {
    /**
     * Prevents user actions on the button
     */
    disabled: PropTypes.bool,
    /**
     * Called when user moves focus away from the button
     */
    onBlur: PropTypes.func,
    /**
     * Called when user clicks the button
     */
    onClick: PropTypes.func,
    /**
     * Called when user moves focus onto the button
     */
    onFocus: PropTypes.func,
    /**
     * Called when mouse is pressed over the button
     */
    onMouseDown: PropTypes.func,
    /**
     * Called when mouse begins to move over the button
     */
    onMouseEnter: PropTypes.func,
    /**
     * Called when mouse stops moving over the button
     */
    onMouseLeave: PropTypes.func,
    /**
     * Called when mouse is released over the button
     */
    onMouseUp: PropTypes.func,
    /**
     * Function to modify the component's styles
     */
    stylesheet: PropTypes.func
  };

  render() {
    const {
      disabled,
      onBlur,
      onFocus,
      onMouseDown,
      onMouseEnter,
      onMouseLeave,
      onMouseUp,
      onClick,
      stylesheet: customStylesheet,
      ...otherProps
    } = this.props;

    function handleClick(event) {
      event.preventDefault();
      event.stopPropagation();
      onClick(event);
    }

    return (
      <ControlBehavior
        onBlur={onBlur}
        onFocus={onFocus}
        onMouseDown={onMouseDown}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onMouseUp={onMouseUp}
      >
        {({
          hasFocus,
          hasHover,
          isPressed,
          onBlur: handleBlur,
          onFocus: handleFocus,
          onMouseDown: handleMouseDown,
          onMouseEnter: handleMouseEnter,
          onMouseLeave: handleMouseLeave,
          onMouseUp: handleMouseUp
        }) => (
          <ThemeContext.Consumer>
            {({ resolvedRoles, metadata }) => {
              const styles = stylesheet(
                { hasFocus, hasHover, isPressed, stylesheet: customStylesheet },
                resolvedRoles,
                metadata
              );
              const { className } = otherProps;
              const CloseIcon =
                metadata.densityId === "medium-density" ? CloseSUI : CloseXsUI;

              return (
                <button
                  onBlur={handleBlur}
                  onFocus={handleFocus}
                  onMouseDown={handleMouseDown}
                  onMouseEnter={handleMouseEnter}
                  onMouseLeave={handleMouseLeave}
                  onMouseUp={handleMouseUp}
                  className={cx(css(styles.button), className)}
                  tabIndex="-1"
                  onClick={handleClick}
                  disabled={disabled}
                  title="close"
                >
                  <CloseIcon />
                </button>
              );
            }}
          </ThemeContext.Consumer>
        )}
      </ControlBehavior>
    );
  }
}
