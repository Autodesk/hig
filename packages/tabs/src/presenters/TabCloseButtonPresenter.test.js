import React from "react";
import renderer from "react-test-renderer";
import TabCloseButtonPresenter from "./TabCloseButtonPresenter";

describe("tabs/TabCloseButtonPresenter", () => {
  const cases = [
    {
      description: "renders without props",
      props: {},
    },
    {
      description: "renders with all props",
      props: {
        disabled: true,
        onBlur: () => {},
        onFocus: () => {},
        onMouseDown: () => {},
        onMouseEnter: () => {},
        onMouseLeave: () => {},
        onMouseUp: () => {},
        onClick: () => {},
      },
    },
  ];

  cases.forEach(({ description, props: { children, ...otherProps } }) => {
    it(description, () => {
      const wrapper = <TabCloseButtonPresenter {...otherProps} />;
      const tree = renderer.create(wrapper).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
