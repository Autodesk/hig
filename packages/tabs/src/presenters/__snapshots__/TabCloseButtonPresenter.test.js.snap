// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tabs/TabCloseButtonPresenter renders with all props 1`] = `
.emotion-1 {
  box-shadow: none;
  background-color: transparent;
  outline: none;
  border: none;
  padding: 0;
  margin: 0;
  width: 16px;
  height: 16px;
  cursor: pointer;
}

.emotion-1 svg * {
  fill: #808080;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  -webkit-transition-property: fill;
  transition-property: fill;
}

.emotion-0 {
  fill: #808080;
  pointer-events: none;
}

.emotion-0 > * {
  fill: #808080;
  pointer-events: none;
}

<button
  className="emotion-1"
  disabled={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  tabIndex="-1"
  title="close"
>
  <svg
    className="emotion-0"
    height="16px"
    viewBox="0 0 16 16"
    width="16px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M13.4 3.4l-.8-.8L8 7.3 3.4 2.6l-.8.8L7.3 8l-4.7 4.6.8.8L8 8.7l4.6 4.7.8-.8L8.7 8l4.7-4.6z"
    />
  </svg>
</button>
`;

exports[`tabs/TabCloseButtonPresenter renders without props 1`] = `
.emotion-1 {
  box-shadow: none;
  background-color: transparent;
  outline: none;
  border: none;
  padding: 0;
  margin: 0;
  width: 16px;
  height: 16px;
  cursor: pointer;
}

.emotion-1 svg * {
  fill: #808080;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  -webkit-transition-property: fill;
  transition-property: fill;
}

.emotion-0 {
  fill: #808080;
  pointer-events: none;
}

.emotion-0 > * {
  fill: #808080;
  pointer-events: none;
}

<button
  className="emotion-1"
  disabled={undefined}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  tabIndex="-1"
  title="close"
>
  <svg
    className="emotion-0"
    height="16px"
    viewBox="0 0 16 16"
    width="16px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M13.4 3.4l-.8-.8L8 7.3 3.4 2.6l-.8.8L7.3 8l-4.7 4.6.8.8L8 8.7l4.6 4.7.8-.8L8.7 8l4.7-4.6z"
    />
  </svg>
</button>
`;
